az-cvm-vtpm/src/vtpm/mod.rs

```diff
@@ -126,6 +126,37 @@ impl Quote {
         let nonce = attest.extra_data().to_vec();
         Ok(nonce)
     }
+
+    /// Extract PCR values for sha256 bank
+    pub fn sha256() -> Result<[String; 24], tss_esapi::Error> {
```

Comment:
If we do a static fn like this this, we wouldn't get a signed quote but re-read unsigned PCR values from the HW. We cannot call this from the verifier (it's not in the TEE) and even if we could, we couldn't verify whether the PCRs are correct and correlate with the quote, I think.

I think the signature should be `pub fn sha256(&self) -> ...` so we extract values from the message, like we do with the nonce.

Reply:
So do you mean the code for this function should live in the `get_quote` ? And I should store the return value of this function in a new field of Quote? And when someone calls `sha256` func it should just return value that is stored earlier?

Reply:
If we just store it alongside the message and signature we cannot prove that the values are correct, because this data is not signed, only the message. so I suspect we have to somehow retrieve it from the message.

Reply:
The API is a bit rough, but I think what we need to do to extract the digests from the signed Quote is this: unmarshall into an `Attest` struct (like we do for extracting the nonce), then get `AttestInfo` via `Attest::attested()`, then pattern match this into a `QuoteInfo`, which holds the digests.

Reply:
We then would calculate the digest of the PCR values (that we added to the evidence) and verify whether it matches the digest in the quote. By doing this we established the PCRs values that we sent as part of the evidence are genuine.

I haven't verified this, but this is how I understand `tpm2_checkquote` to work, see [here](https://github.com/tpm2-software/tpm2-tools/blob/beb8d63171a6f3e05371fe8ce549ef46161b0550/tools/misc/tpm2_checkquote.c#L493C31-L493C31) and [here](https://github.com/tpm2-software/tpm2-tools/blob/beb8d63171a6f3e05371fe8ce549ef46161b0550/tools/misc/tpm2_checkquote.c#L168)

Reply:
> If we just store it alongside the message and signature we cannot prove that the values are correct, because this data is not signed, only the message. so I suspect we have to somehow link it to the message.

Reply:
I verified this now on a CVM. it seems to work, if you calculate a SHA256 digest from the requested PCRs, it is supposed to match the digest in the quote:

```rust
pub fn verify_pcrs(&self) -> Result<(), VerifyError> {
    let mut hasher = Sha256::new();
    for pcr in self.pcrs.iter() {
        hasher.update(pcr);
    }
    let digest = hasher.finalize();
    let attest = Attest::unmarshall(&self.message)?;
    let AttestInfo::Quote { info } = attest.attested() else {
        panic!("QuoteInfo not found");
    };

    let quote_digest = info.pcr_digest();
    if digest[..] != quote_digest[..] {
        panic!("digest doesn't match: {:?} != {:?}", quote_digest, digest);
    }

    Ok(())
}
```

Reply:
So, we are just verifying and not sending the actual hashes to be able to be used with RVPS?

Also, the name `verify_pcrs` (plural) is misleading, isn't it? We are only checking on PCR and not all, isn't it? Because the method `info.pcr_digest()` returns only one PCR's value, right?

Also who populates the `self.pcrs` field?

Reply:
> So, we are just verifying and not sending the actual hashes to be able to be used with RVPS

No, it's reasonable to send the full PCR bank (we just support sha256 atm) as part of the evidence, so we can check each item individually.

But we _also_ need to verify that the PCRs we send are legit and not some arbitrary values, this is what `verify_pcrs()` would do. it checks whether the digest of the PCRs is equal to the digest in the TPM quote.

> Also, the name verify_pcrs (plural) is misleading, isn't it? We are only checking on PCR and not all, isn't it? Because the method info.pcr_digest() returns only one PCR's value, right?

We would verify multiple PCRs. The digest that `info.pcr_digest()` returns is a single digest of all the PCR values that were [picked as part of the quote](https://github.com/kinvolk/azure-cvm-tooling/blob/b353d341bbcfc032ff4d5d20f694c6cee290a7f0/az-cvm-vtpm/src/vtpm/mod.rs#L148-L150). a hash of hashes if you will. so we replay the hashing and compare the result, to establish that the PCRs values are the same that have been used in the quote.

> Also who populates the self.pcrs field?

That would be us, we can add a field to the `Quote` struct and populate it using the `pcr::read_all()` fn, probably as part of a `get_quote()` call.

Reply:
@mkulke PTAL.